File: python/examples/type_system.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# type_system: Type system overview -- structs, aliases, enums, and full lifecycle.
#
# Usage: python type_system.py [host_url]
#
# Defaults: http://127.0.0.1:18080, expects a program already open in Ghidra.

import sys

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. List existing types
    try:
        types_resp = client.list_types(limit=10)
    except ghidra.GhidraError as e:
        print(f"list_types failed: {e}", file=sys.stderr)
        sys.exit(1)

    types = types_resp.types
    print(f"Existing types ({len(types)} shown):")
    for t in types:
        defined = "" if not t.is_not_yet_defined else "  [undefined]"
        print(f"  {t.name:<28}  kind={t.kind:<10}  size={t.length:>4}{defined}")

    # 2. Get details on a specific type
    if types:
        sample = types[0]
        print(f"\nDetails for '{sample.name}':")
        try:
            detail = client.get_type(sample.path_name)
            if detail.type:
                t = detail.type
                print(f"  id={t.type_id}  path={t.path_name}  category={t.category_path}")
                print(f"  display={t.display_name}  kind={t.kind}  length={t.length}")
                if t.source_archive:
                    print(f"  source_archive={t.source_archive}")
        except ghidra.GhidraError as e:
            print(f"  get_type failed: {e}", file=sys.stderr)

    # 3. Create a struct type
    struct_name = "ExampleStruct"
    print(f"\nCreating struct '{struct_name}' (size=16)...")
    try:
        create_resp = client.create_type(struct_name, "struct", 16)
        print(f"  updated={create_resp.updated}")
    except ghidra.GhidraError as e:
        print(f"  create_type failed: {e}", file=sys.stderr)
        sys.exit(1)

    # Verify creation
    verify = client.get_type(f"/{struct_name}")
    if verify.type:
        print(f"  Verified: {verify.type.name} (kind={verify.type.kind}, size={verify.type.length})")

    # 4. Create a type alias pointing to our struct
    alias_name = "ExampleAlias"
    print(f"\nCreating alias '{alias_name}' -> '{struct_name}'...")
    try:
        alias_resp = client.create_type_alias(alias_name, struct_name)
        print(f"  updated={alias_resp.updated}")
    except ghidra.GhidraError as e:
        print(f"  create_type_alias failed: {e}", file=sys.stderr)

    # 5. Create an enum type
    enum_name = "ExampleEnum"
    print(f"\nCreating enum '{enum_name}' (width=4)...")
    try:
        enum_resp = client.create_type_enum(enum_name, width=4, is_signed=False)
        print(f"  updated={enum_resp.updated}")
    except ghidra.GhidraError as e:
        print(f"  create_type_enum failed: {e}", file=sys.stderr)

    # 6. Rename the struct type
    renamed = "ExampleStructRenamed"
    print(f"\nRenaming '{struct_name}' -> '{renamed}'...")
    try:
        ren_resp = client.rename_type(f"/{struct_name}", renamed)
        print(f"  updated={ren_resp.updated}  name={ren_resp.name}")
    except ghidra.GhidraError as e:
        print(f"  rename_type failed: {e}", file=sys.stderr)
        renamed = struct_name  # fall back for cleanup

    # 7. List aliases, unions, and enums
    print("\nType aliases:")
    try:
        aliases = client.list_type_aliases(limit=10)
        for a in aliases.aliases:
            print(f"  {a.name}")
        if not aliases.aliases:
            print("  (none)")
    except ghidra.GhidraError as e:
        print(f"  list_type_aliases failed: {e}", file=sys.stderr)

    print("\nType unions:")
    try:
        unions = client.list_type_unions(limit=10)
        for u in unions.unions:
            print(f"  {u.name}")
        if not unions.unions:
            print("  (none)")
    except ghidra.GhidraError as e:
        print(f"  list_type_unions failed: {e}", file=sys.stderr)

    print("\nType enums:")
    try:
        enums = client.list_type_enums(limit=10)
        for en in enums.enums:
            print(f"  {en.name}")
        if not enums.enums:
            print("  (none)")
    except ghidra.GhidraError as e:
        print(f"  list_type_enums failed: {e}", file=sys.stderr)

    # 8. Clean up -- delete all created types (reverse order of dependencies)
    print("\nCleaning up...")

    # Delete alias first (depends on struct)
    try:
        client.delete_type_alias(f"/{alias_name}")
        print(f"  Deleted alias '{alias_name}'")
    except ghidra.GhidraError as e:
        print(f"  delete alias failed: {e}", file=sys.stderr)

    # Delete enum
    try:
        client.delete_type_enum(f"/{enum_name}")
        print(f"  Deleted enum '{enum_name}'")
    except ghidra.GhidraError as e:
        print(f"  delete enum failed: {e}", file=sys.stderr)

    # Delete struct (renamed)
    try:
        client.delete_type(f"/{renamed}")
        print(f"  Deleted struct '{renamed}'")
    except ghidra.GhidraError as e:
        print(f"  delete struct failed: {e}", file=sys.stderr)

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/type_system.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// type_system: Browse types, create structs, aliases, enums, and clean up.
//
// Usage: type_system [host_url]

use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. List existing types (first page)
    let types = match client.list_types("", 15, 0) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("ListTypes failed (is a program open?): {e}");
            std::process::exit(1);
        }
    };

    println!("Types ({} shown):", types.types.len());
    for t in &types.types {
        println!(
            "  {:<30} kind={:<10} len={:<4} path={}",
            t.name, t.kind, t.length, t.path_name
        );
    }

    // 2. Search for a specific built-in type
    println!("\nSearching for 'int' types...");
    let int_types = client.list_types("int", 10, 0).unwrap();
    for t in &int_types.types {
        println!("  {:<20} kind={}", t.name, t.kind);
    }

    // 3. Get details for a specific type
    if let Some(first) = types.types.first() {
        let detail = client.get_type(&first.path_name).unwrap_or_else(|e| {
            eprintln!("GetType failed: {e}");
            std::process::exit(1);
        });
        if let Some(t) = &detail.r#type {
            println!("\nType detail for '{}':", t.name);
            println!("  ID:       {}", t.type_id);
            println!("  Path:     {}", t.path_name);
            println!("  Category: {}", t.category_path);
            println!("  Display:  {}", t.display_name);
            println!("  Kind:     {}", t.kind);
            println!("  Length:   {}", t.length);
        }
    }

    // 4. Create a new struct type
    let struct_name = "RustExampleStruct";
    println!("\nCreating struct '{}'...", struct_name);
    let create = client
        .create_type(struct_name, "struct", 16)
        .unwrap_or_else(|e| {
            eprintln!("CreateType failed: {e}");
            std::process::exit(1);
        });
    println!("  updated={}", create.updated);

    // 5. Verify the struct exists
    let search = client.list_types(struct_name, 5, 0).unwrap();
    println!(
        "  Found {} type(s) matching '{}'",
        search.types.len(),
        struct_name
    );

    // 6. Create a type alias pointing to our struct
    let alias_name = "RustExampleAlias";
    println!("\nCreating alias '{}' -> '{}'...", alias_name, struct_name);
    let alias = client
        .create_type_alias(alias_name, struct_name)
        .unwrap_or_else(|e| {
            eprintln!("CreateTypeAlias failed: {e}");
            std::process::exit(1);
        });
    println!("  updated={}", alias.updated);

    // 7. List aliases to verify
    let aliases = client.list_type_aliases(alias_name, 5, 0).unwrap();
    println!(
        "  Aliases matching '{}': {}",
        alias_name,
        aliases.aliases.len()
    );

    // 8. Create an enum type
    let enum_name = "RustExampleEnum";
    println!("\nCreating enum '{}' (width=4, signed=false)...", enum_name);
    let enu = client
        .create_type_enum(enum_name, 4, false)
        .unwrap_or_else(|e| {
            eprintln!("CreateTypeEnum failed: {e}");
            std::process::exit(1);
        });
    println!("  updated={}", enu.updated);

    // 9. Rename the struct
    let new_struct_name = "RustExampleStructRenamed";
    println!("\nRenaming '{}' -> '{}'...", struct_name, new_struct_name);
    let rn = client
        .rename_type(struct_name, new_struct_name)
        .unwrap_or_else(|e| {
            eprintln!("RenameType failed: {e}");
            std::process::exit(1);
        });
    println!("  updated={}, name=\"{}\"", rn.updated, rn.name);

    // 10. Clean up: delete everything we created
    println!("\nCleaning up...");

    let del_alias = client.delete_type_alias(alias_name);
    println!(
        "  Delete alias '{}': {}",
        alias_name,
        match &del_alias {
            Ok(d) => format!("deleted={}", d.deleted),
            Err(e) => format!("error={e}"),
        }
    );

    let del_enum = client.delete_type_enum(enum_name);
    println!(
        "  Delete enum '{}': {}",
        enum_name,
        match &del_enum {
            Ok(d) => format!("deleted={}", d.deleted),
            Err(e) => format!("error={e}"),
        }
    );

    let del_struct = client.delete_type(new_struct_name);
    println!(
        "  Delete struct '{}': {}",
        new_struct_name,
        match &del_struct {
            Ok(d) => format!("deleted={}", d.deleted),
            Err(e) => format!("error={e}"),
        }
    );

    // 11. Confirm cleanup
    let final_check = client.list_types("RustExample", 10, 0).unwrap();
    println!(
        "\nTypes matching 'RustExample' after cleanup: {}",
        final_check.types.len()
    );

    println!("\nDone.");
}
